File: whatsapp-agent/src/contentFilter.js
'use strict';

// JS port of backend/services/content_filter.py
// IMPORTANT: Do NOT add the 'g' flag to patterns used with .test() — it would
// advance lastIndex and cause alternating match/no-match on repeated calls.

const RISK_PATTERNS = [
    {
        level: 'high',
        patterns: [
            /\b(çocuk\s*porn|child\s*porn|cp\s+link|pedofil|lolita)\b/iu,
            /\b(tecavüz\s*video|rape\s*video|snuff\s*film)\b/iu,
            /\b(uyuşturucu\s*sat|eroin|kokain|metamfetamin|bonzai\s*sat|esrar\s*sat)\b/iu,
        ],
    },
    {
        level: 'medium',
        patterns: [
            /\b(seks\s*video|sex\s*tape|nude\s*photo|çıplak\s*foto|sikişme)\b/iu,
            /\b(onlyfans\.com|pornhub\.com|xvideos\.com|xnxx\.com|redtube\.com|xhamster\.com)\b/iu,
            /\b(webcam\s*show|cam\s*girl|escort\s*ilan|masaj\s*escort)\b/iu,
            /\b(kumar\s*oyna|bahis\s*oyna|slot\s*oyna|casino\s*kazan|iddaa\s*tahmin)\b/iu,
        ],
    },
    {
        level: 'low',
        patterns: [
            /\b(gizlice\s*buluş|gizli\s*buluşma|randevu\s*atalım\s*kimse\s*bilmesin)\b/iu,
            /\b(tinder|badoo|grindr|bumble|happn|blendr)\b/iu,
            /\b(fotoğrafını\s*gönder|resim\s*at\s*bana|body\s*pic|dick\s*pic|sexy\s*photo|çıplak\s*gönder)\b/iu,
            /\b(ailene\s*söyleme|kimseye\s*söyleme|sır\s*tutalım|bunu\s*bilmesin|annene\s*söyleme)\b/iu,
        ],
    },
];

const LEVEL_ORDER = { none: 0, low: 1, medium: 2, high: 3 };
const BASE_SCORES  = { none: 0, low: 15, medium: 35, high: 60 };

function classifyRisk(text) {
    if (!text) return { riskLevel: 'none', riskCategories: [], riskScore: 0 };

    let highestLevel = 'none';
    const matchedCategories = new Set();

    for (const { level, patterns } of RISK_PATTERNS) {
        for (const pattern of patterns) {
            if (pattern.test(text)) {
                matchedCategories.add(level);
                if (LEVEL_ORDER[level] > LEVEL_ORDER[highestLevel]) {
                    highestLevel = level;
                }
                break;
            }
        }
    }

    return {
        riskLevel: highestLevel,
        riskCategories: [...matchedCategories],
        riskScore: 0, // caller computes final score with deletion flags
    };
}

function calcRiskScore(riskLevel, isDeleted, hasMedia, isFromMe) {
    let score = BASE_SCORES[riskLevel] ?? 0;
    if (isDeleted && hasMedia)      score += 30;
    else if (isDeleted && isFromMe) score += 25;
    else if (isDeleted)             score += 15;
    return Math.min(score, 100);
}

// ── Financial redaction ──────────────────────────────────────────────────────

const CARD_RE   = /\b(\d{4}[\s-]?\d{4}[\s-]?\d{4}[\s-]?\d{4})\b/g;
const CVV_RE    = /(cvv|cvc|güvenlik\s*kod[iu]|security\s*code)\s*[:\-]?\s*(\d{3,4})\b/gi;
const PIN_RE    = /(kart\s*şifre(?:si|m|n)?|pin\s*kod(?:u|um|un)?|atm\s*şifre(?:si|m|n)?)\s*[:\-]?\s*(\d{4,6})\b/gi;
const OTP_RE    = /(doğrulama\s*kod[iu]?|verification\s*code|otp|tek\s*kullanım)\s*[:\-]?\s*(\d{4,8})\b/gi;
const CRYPTO_RE = /(seed\s*phrase|mnemonic|private\s*key|cüzdan\s*şifre|wallet\s*key)\s*[:\-]?\s*\S+/gi;
const IBAN_RE   = /\b([A-Z]{2}\d{2}[\d\s]{10,30})\b/g;

function redactText(text) {
    if (!text) return { redactedText: text, wasRedacted: false };

    let t = text;
    let modified = false;

    // Preserve IBANs via placeholders so card regex doesn't clobber them
    const ibanMap = {};
    let ibanIdx = 0;
    t = t.replace(IBAN_RE, (m) => {
        const key = `__IBAN${ibanIdx++}__`;
        ibanMap[key] = m;
        return key;
    });

    function rep(re, fn) {
        const result = t.replace(re, fn);
        if (result !== t) modified = true;
        t = result;
    }

    rep(CARD_RE,   (_, num) => {
        const digits = num.replace(/[\s-]/g, '');
        return `**** **** **** ${digits.slice(-4)}`;
    });
    rep(CVV_RE,    (_, label) => `${label}: [GİZLİ]`);
    rep(PIN_RE,    (_, label) => `${label}: [GİZLİ]`);
    rep(OTP_RE,    (_, label) => `${label}: [GİZLİ]`);
    rep(CRYPTO_RE, ()         => '[KRİPTO BİLGİSİ GİZLENDİ]');

    // Restore IBANs
    for (const [key, val] of Object.entries(ibanMap)) {
        t = t.replace(key, val);
    }

    return { redactedText: t, wasRedacted: modified };
}

module.exports = { classifyRisk, calcRiskScore, redactText };

File: backend/services/content_filter.py
import re
from enum import Enum
from typing import Tuple, List

_CARD_NUMBER_RE = re.compile(r'\b(\d{4}[\s\-]?\d{4}[\s\-]?\d{4}[\s\-]?\d{4})\b')
_CVV_RE = re.compile(r'(?i)(cvv|cvc|güvenlik\s*kod[iu]|security\s*code)\s*[:\-]?\s*(\d{3,4})\b')
_PIN_RE = re.compile(r'(?i)(kart\s*şifre(?:si|m|n)?|pin\s*kod(?:u|um|un)?|atm\s*şifre(?:si|m|n)?)\s*[:\-]?\s*(\d{4,6})\b')
_IBAN_RE = re.compile(r'\b([A-Z]{2}\d{2}[\d\s]{10,30})\b')
_OTP_RE = re.compile(r'(?i)(doğrulama\s*kod[iu]?|verification\s*code|otp|tek\s*kullanım)\s*[:\-]?\s*(\d{4,8})\b')
_CRYPTO_RE = re.compile(r'(?i)(seed\s*phrase|mnemonic|private\s*key|cüzdan\s*şifre|wallet\s*key)\s*[:\-]?\s*\w+')

_CARD_IMAGE_KEYWORDS = ['kart','card','visa','mastercard','troy','amex','banka','bank','kredi','credit','debit']

# Finansal uygulama bundle ID'leri — bu uygulamaların ekran görüntüleri toplanmaz
BLOCKED_FINANCIAL_APPS = [
    'com.garanti.cepsubesi', 'com.isbankasi.iscepcanlisi',
    'com.akbank.softo', 'com.ziraatbankasi.ziraatmobil',
    'com.yapikredi.mobile', 'com.halkbank.mobile',
    'com.binance.exchange', 'com.btcturk.app', 'com.paribu.app',
    'com.agileBits.onepassword', 'com.apple.keychainaccess',
]


def redact_text(text: str) -> Tuple[str, bool]:
    """Hassas finansal bilgileri redakte eder. IBAN'a dokunmaz."""
    modified = False
    iban_placeholders = {}

    def save_iban(m):
        key = f"__IBAN_{len(iban_placeholders)}__"
        iban_placeholders[key] = m.group(0)
        return key

    text = _IBAN_RE.sub(save_iban, text)

    def redact_card(m):
        full = m.group(1).replace(' ', '').replace('-', '')
        return f"**** **** **** {full[-4:]}"

    new_text = _CARD_NUMBER_RE.sub(redact_card, text)
    if new_text != text: modified = True
    text = new_text

    def redact_cvv(m): return f"{m.group(1)}: [GİZLİ]"
    new_text = _CVV_RE.sub(redact_cvv, text)
    if new_text != text: modified = True
    text = new_text

    def redact_pin(m): return f"{m.group(1)}: [GİZLİ]"
    new_text = _PIN_RE.sub(redact_pin, text)
    if new_text != text: modified = True
    text = new_text

    def redact_otp(m): return f"{m.group(1)}: [GİZLİ]"
    new_text = _OTP_RE.sub(redact_otp, text)
    if new_text != text: modified = True
    text = new_text

    def redact_crypto(m): return "[KRİPTO BİLGİSİ GİZLENDİ]"
    new_text = _CRYPTO_RE.sub(redact_crypto, text)
    if new_text != text: modified = True
    text = new_text

    for key, value in iban_placeholders.items():
        text = text.replace(key, value)

    return text, modified


def is_card_image(filename: str, metadata: dict = None) -> bool:
    name_lower = filename.lower()
    for kw in _CARD_IMAGE_KEYWORDS:
        if kw in name_lower: return True
    if metadata:
        description = str(metadata).lower()
        for kw in _CARD_IMAGE_KEYWORDS:
            if kw in description: return True
    return False


def is_blocked_app(bundle_id: str) -> bool:
    """Finansal/hassas uygulama mı kontrol eder."""
    return bundle_id.lower() in [b.lower() for b in BLOCKED_FINANCIAL_APPS]


class RiskLevel(str, Enum):
    NONE = "none"
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"


_RISK_PATTERNS: List[Tuple[RiskLevel, List[str]]] = [
    (RiskLevel.HIGH, [
        r'\b(çocuk\s*porn|child\s*porn|cp\s+link|pedofil|lolita)\b',
        r'\b(tecavüz|rape\s+video|snuff)\b',
        r'\b(uyuşturucu|eroin|kokain|metamfetamin|uyuşturucu\s*sat)\b',
    ]),
    (RiskLevel.MEDIUM, [
        r'\b(seks\s*video|sex\s*tape|nude\s*photo|çıplak\s*foto|sikişme|am\s+ver|got\s+ver)\b',
        r'\b(onlyfans\.com|pornhub\.com|xvideos\.com|xnxx\.com|redtube\.com)\b',
        r'\b(webcam\s*show|cam\s*girl|escort\s*ilan|masaj\s*escort)\b',
        r'\b(kumar|bahis\s*oyna|slot\s*oyna|casino\s*kazan)\b',
    ]),
    (RiskLevel.LOW, [
        r'\b(buluşalım\s*mı\s*gizlice|sana\s*gel\s*mi|gizli\s*buluşma|randevu\s*atalım)\b',
        r'\b(tinder|badoo|grindr|bumble|happn)\b',
        r'\b(fotoğrafını\s*gönder|resim\s*at\s*bana|body\s*pic|dick\s*pic|sexy\s*photo)\b',
        r'\b(ailene\s*söyleme|kimseye\s*söyleme|sır\s*tutalım|bunu\s*bilmesin)\b',
    ]),
]

_COMPILED_PATTERNS = [
    (level, [re.compile(p, re.IGNORECASE | re.UNICODE) for p in patterns])
    for level, patterns in _RISK_PATTERNS
]


def classify_risk(text: str) -> Tuple[RiskLevel, List[str]]:
    if not text: return RiskLevel.NONE, []
    matched_categories: List[str] = []
    highest_level = RiskLevel.NONE
    level_order = {RiskLevel.NONE: 0, RiskLevel.LOW: 1, RiskLevel.MEDIUM: 2, RiskLevel.HIGH: 3}
    for level, patterns in _COMPILED_PATTERNS:
        for pattern in patterns:
            if pattern.search(text):
                matched_categories.append(level.value)
                if level_order[level] > level_order[highest_level]:
                    highest_level = level
                break
    return highest_level, list(set(matched_categories))


def analyze_whatsapp_message(text: str) -> Tuple[str, bool, RiskLevel, List[str]]:
    filtered_text, was_redacted = redact_text(text)
    risk_level, risk_categories = classify_risk(text)
    return filtered_text, was_redacted, risk_level, risk_categories
